src/NUnitFramework/framework/Internal/MessagePumpStrategy.cs

```diff
@@ -0,0 +1,136 @@
+// ***********************************************************************
+// Copyright (c) 2018 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+#if ASYNC
+using System;
+using System.Threading;
+
+#if NET40 || NET45
+using System.Windows.Forms;
+using System.Windows.Threading;
+#endif
+
+namespace NUnit.Framework.Internal
+{
+    internal abstract class MessagePumpStrategy
+    {
+        public abstract void WaitForCompletion(AwaitAdapter awaitable);
+
+        public static MessagePumpStrategy FromCurrentSynchronizationContext()
+        {
+#if NET40 || NET45
+            var context = SynchronizationContext.Current;
+
+            if (context is WindowsFormsSynchronizationContext)
+                return WindowsFormsMessagePumpStrategy.Instance;
+
+            if (context is DispatcherSynchronizationContext)
```

Comment:
Will we need to be adding additional message pump strategies for other platforms like Xamarin, possibly Unity?

Reply:
If so, do we need issues to track these or do we wait for user requests?

Reply:
@rprouse I don't know enough about Xamarin and Unity to know whether they run message pumps.
https://developer.xamarin.com/api/type/Android.OS.Looper/ suggests that they do, but unless we can test these in CI, I'm not inclined to do any work on them. I spent an hour trying to figure out why VS starts the Visual Studio Android emulator but never finishes deploying a Xamarin Android project to it, so suffice it to say that I'm not in great shape for that on my main machine.

Reply:
\<rant>
The state of Android emulation on Windows is unreasonable. HAXM is no good because it requires me to shut down my Hyper-V virtual machines, one of which is a server, not to mention Docker. ARM is ten times slower since it's not hardware accelerated. The Visual Studio Android Emulator is perfect because it's accelerated and works without disabling Hyper-V, but I've never gotten it to work on my new machine and they only have up to API level 23 images.
\</rant>

Reply:
😍 https://blogs.msdn.microsoft.com/visualstudio/2018/05/08/hyper-v-android-emulator-support/